### PR TITLE
Simplify IndexShard indexing / deletion methods

### PR DESCRIPTION
**Pull request by ywelsch**
Indexing or deleting documents through the `IndexShard` interface is quite complex and error-prone. It requires multiple calls, e.g. first `prepareIndexOnPrimary`, then do some checks if mapping updates have occurred, then do the actual indexing using `index(...)` etc. Currently each consumer of the interface (local recovery, peer recovery, replication) has additional custom checks built around it to deal with mapping updates, some of which are even inconsistent (cf. https://github.com/elastic/elasticsearch/pull/24858#discussion_r120304497). This PR aims at reducing the complexity by exposing a simpler interface on IndexShard. There are no more `prepare***` methods and the mapping complexity is also hidden, but still giving callers a possibility to implement custom logic to deal with mapping updates.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/index/shard/IndexShard.java [hunk 1]
```diff
@@ -628,32 +607,65 @@ private IndexShardState changeState(IndexShardState newState, String reason) {
...
+        if (indexSettings().isSingleType()) {
+            // When there is a single type, the unique identifier is only composed of the _id,
+            // so there is no way to differenciate foo#1 from bar#1. This is especially an issue
+            // if a user first deletes foo#1 and then indexes bar#1: since we do not encode the
+            // _type in the uid it might look like we are reindexing the same document, which
+            // would fail if bar#1 is indexed with a lower version than foo#1 was deleted with.
+            // In order to work around this issue, we make deletions create types. This way, we
+            // fail if index and delete operations do not use the same type.
+            try {
+                Mapping update = mapperService().documentMapperWithAutoCreate(type).getMapping();
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: shall we use docMapper(type) to be consistent with applyIndexOperation?

**Review by bleskes — APPROVED** [review]
This is great. I left some nits and one follow discussion.

File: core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java [hunk 2]
```diff
@@ -477,12 +463,12 @@ static ReplicaItemExecutionMode replicaItemExecutionMode(final BulkItemRequest r
                     case FAILURE:
                         final BulkItemResponse.Failure failure = item.getPrimaryResponse().getFailure();
                         assert failure.getSeqNo() != SequenceNumbersService.UNASSIGNED_SEQ_NO : "seq no must be assigned";
-                        operationResult = executeFailureNoOpOnReplica(failure, primaryTerm, replica);
+                        operationResult = replica.applyNoOpOnReplica(failure.getSeqNo(), primaryTerm, failure.getMessage());
```
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java]: addressed in 760a024

File: core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java [hunk 3]
```diff
@@ -495,6 +481,37 @@ static ReplicaItemExecutionMode replicaItemExecutionMode(final BulkItemRequest r
         return location;
     }
 
+    private static Engine.Result performNormalOpOnReplica(DocWriteResponse primaryResponse, DocWriteRequest docWriteRequest,
```
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java]: 760a024

File: core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java [hunk 4]
```diff
@@ -527,172 +544,49 @@ static ReplicaItemExecutionMode replicaItemExecutionMode(final BulkItemRequest r
...
-            mappingUpdateNeeded = mappingUpdate != null;
-            if (mappingUpdateNeeded) {
-                mappingUpdater.updateMappings(mappingUpdate, primary.shardId(), request.type());
-            }
-        } catch (MapperParsingException | IllegalArgumentException failure) {
-            return new Engine.IndexResult(failure, request.version());
+            return primary.applyIndexOperationOnPrimary(request.version(), request.versionType(), sourceToParse,
+                request.getAutoGeneratedTimestamp(), request.isRetry(), update -> {
+                    mappingUpdater.updateMappings(update, primary.shardId(), sourceToParse.type());
+                    throw new ReplicationOperation.RetryOnPrimaryException(primary.shardId(), "Mapping updated");
```
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java]: I've added a comment and a TODO how we could possibly simplify this. I think this needs a bit more study though, and leave that to a future PR.

File: core/src/main/java/org/elasticsearch/index/shard/IndexShard.java [hunk 5]
```diff
@@ -531,34 +535,46 @@ private IndexShardState changeState(IndexShardState newState, String reason) {
...
-            assert opPrimaryTerm <= this.primaryTerm : "op term [ " + opPrimaryTerm + " ] > shard term [" + this.primaryTerm + "]";
-            return prepareIndex(docMapper(source.type()), source, opSeqNo, opPrimaryTerm, version, versionType,
-                    Engine.Operation.Origin.REPLICA, autoGeneratedIdTimestamp, isRetry);
+            operation = prepareIndex(docMapper(sourceToParse.type()), sourceToParse, seqNo, opPrimaryTerm, version, versionType, origin,
+                autoGeneratedTimeStamp, isRetry);
+            Mapping update = operation.parsedDoc().dynamicMappingsUpdate();
+            if (update != null) {
+                onMappingUpdate.accept(update);
+            }
+        } catch (MapperParsingException | IllegalArgumentException e) {
```
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: added a comment, and also added handling for TypeMissingException (+ test)

File: core/src/main/java/org/elasticsearch/index/shard/IndexShard.java (hunk 1 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: sure

**Comment by ywelsch** [comment]
thanks @bleskes

**Pull request merged**